contracts/FeralfileArtworkV4.sol

```diff
@@ -1,112 +1,602 @@
+    /// @param seriesId a series ID
+    /// @return uint256 the total supply
+    function seriesTotalSupply(
+        uint256 seriesId
+    ) external view virtual returns (uint256) {
+        return _seriesTotalSupplies[seriesId];
+    }
+
+    /// @notice Get artwork data
+    /// @param tokenId a token ID representing the artwork
+    /// @return Artwork the Artwork object
+    function getArtwork(
+        uint256 tokenId
+    ) external view virtual returns (Artwork memory) {
+        require(_exists(tokenId), "ERC721: invalid token ID");
+        return _allArtworks[tokenId];
+    }
+
+    /// @notice Set vault contract
+    /// @dev don't allow to set signer as zero address
+    function setVault(address vault_) external onlyOwner {
+        require(
+            vault_ != address(0),
+            "FeralfileExhibitionV4: vault_ is zero address"
+        );
+        vault = IFeralfileVault(payable(vault_));
+    }
+
+    /// @notice Return flag _selling;
+    function selling() external view returns (bool) {
+        return _selling;
+    }
+
+    function _checkContractOwnedToken() private view {
+        uint256 balance = balanceOf(address(this));
+        require(
+            balance > 0,
+            "FeralfileExhibitionV4: No token owned by the contract"
+        );
+    }
+
+    /// @notice Start token selling
+    function startSale() external onlyOwner {
+        mintable = false;
+        resumeSale();
+    }
+
+    /// @notice Resume token selling
+    function resumeSale() public onlyOwner {
+        require(
+            !mintable,
+            "FeralfileExhibitionV4: mintable required to be false"
+        );
+        require(
+            !_selling,
+            "FeralfileExhibitionV4: _selling required to be false"
+        );
+        _checkContractOwnedToken();
+
+        _selling = true;
+    }
+
+    /// @notice Pause token sale
+    function pauseSale() public onlyOwner {
+        require(
+            !mintable,
+            "FeralfileExhibitionV4: mintable required to be false"
+        );
+        require(
+            _selling,
+            "FeralfileExhibitionV4: _selling required to be true"
+        );
+        _selling = false;
+    }
+
+    /// @notice Stop token selling and burn remaining tokens
+    function stopSaleAndBurn() external onlyOwner {
+        pauseSale();
+
+        // burn remaining tokens
+        uint256[] memory tokenIds = _ownedTokens[address(this)];
+        for (uint256 i = 0; i < tokenIds.length; i++) {
+            _burnArtwork(tokenIds[i]);
+        }
     }
 
-    function supportsInterface(
-        bytes4 interfaceId
-    ) public view virtual override(ERC721Enumerable, IERC165) returns (bool) {
-        return
-            interfaceId == type(IERC721Enumerable).interfaceId ||
-            super.supportsInterface(interfaceId);
+    /// @notice Stop token selling and transfer remaining tokens back to the underlying addresses
+    function stopSaleAndTransfer(
```

Comment:
There is a chance to stop the sale with only part of series transfer tokens to addresses. This means the sale can start with the remaining token again. I think we need to ensure either:
- the length of `seriesIds` is same to the length of `_seriesMaxSupplies`
- balanceOf(address(this)) == 0

---

- The first approach would need to introduce another key since we can not count the length of a map directly.
- The later one would waste more gas since we can only check after we burn all tokens.